src/flamenco/txn/fd_txn_generate.c

```diff
@@ -0,0 +1,160 @@
+#include "fd_txn_generate.h"
+
+/* Message header type */
+struct __attribute__((packed)) fd_txn_message_hdr {
+  uchar num_signatures;
+  uchar num_readonly_signatures;
+  uchar num_readonly_unsigned;
+};
+
+typedef struct fd_txn_message_hdr fd_txn_message_hdr_t;
+
+static fd_txn_instr_t *
+fd_txn_instr_meta_generate( uchar * out_buf,
+                            uchar program_id,
+                            ushort acct_cnt,
+                            ushort data_sz,
+                            ushort acct_off,
+                            ushort data_off ) {
+    fd_txn_instr_t * out_instr = (fd_txn_instr_t *) out_buf;
+    out_instr->program_id = program_id;
+    out_instr->acct_cnt   = acct_cnt;
+    out_instr->data_sz    = data_sz;
+    out_instr->acct_off   = acct_off;
+    out_instr->data_off   = data_off;
+    return out_instr;
+}
+
+ulong
+fd_txn_base_generate( uchar out_txn_meta[ static FD_TXN_MAX_SZ ],
+                      uchar out_txn_payload[ static FD_TXN_MTU ],
+                      ulong num_signatures,
+                      fd_txn_accounts_t * accounts,
+                      uchar * opt_recent_blockhash ) {
+
+  /* Number of signatures cannot exceed 127. */
+  FD_TEST(num_signatures <= FD_TXN_SIG_MAX);
+  *out_txn_payload = (uchar)num_signatures;
+
+  /* Fill out txn metadata */
+  fd_txn_t * txn_meta             = (fd_txn_t *) out_txn_meta;
+  txn_meta->acct_addr_cnt         = accounts->acct_cnt;
+  txn_meta->readonly_signed_cnt   = accounts->readonly_signed_cnt;
+  txn_meta->readonly_unsigned_cnt = accounts->readonly_unsigned_cnt;
+  /* Number of signatures is encoded as a compact u16 but
+     can always be encoded using 1 byte here. */
+  txn_meta->message_off           = (ushort)(num_signatures * FD_TXN_SIGNATURE_SZ  + 1);
```

Comment:
If num_signatures is higher than 12, then this offset is already going beyond `FD_TXN_MTU`. It seems useful to guarantee that this function would never write beyond `FD_TXN_MTU` bytes from `out_txn_payload`.

Reply:
We have it setup such that the caller can specify a larger buffer. That can be changed in a follow up PR if you'd like to take a look at that